migrate from JDBC to JPA

ihavedone opened the pull request:
En: migrate from JDBC to JPA
Ru: миграция с JDBC на JPA

ihavedone committed:
migrate from JDBC to JPA

stvort commented on library/src/main/java/ru/otus/merets/library/domain/Book.java:
```diff
@@ -2,17 +2,42 @@
+    @JoinTable(name= "books_genres", joinColumns = @JoinColumn(name="book_id"),inverseJoinColumns = @JoinColumn(name="genre_id"))
+    private Set<Genre> genres;
+
+    @BatchSize(size = 20)
+    @OneToMany(targetEntity = Comment.class, fetch = FetchType.LAZY, cascade = CascadeType.ALL)
+    @JoinColumn(name = "book_id")
+    List<Comment> comments;
```
Не уверен, что комментарии это свойство книги, скорее наоборот, информация о книге это свойство комментария

stvort commented on library/src/main/java/ru/otus/merets/library/domain/Comment.java:
```diff
@@ -0,0 +1,25 @@
+    @Column(name = "text")
+    private String text;
+
+    @Column(name="book_id")
+    private Long book_id;
```
Лучше прямо книга чтобы

stvort commented on library/src/main/java/ru/otus/merets/library/service/BookServiceImpl.java:
```diff
@@ -1,92 +1,119 @@
     }
 
     @Override
+    @Transactional
```
`readOnly=true` если транзакция только для чтения

ihavedone committed:
en: move comment entity (book is a property of comment, not vice versa), create a comment service (extract from book service)
ru: изменены сущности (комментарий имеет линк на книгу, а не книга содержит комментарии), создан сервис для работы с комментариями

ihavedone committed:
en: added param readOnly for @Transactional annotation where use only for reading
ru: добавил параметры readOnly для транзакций на чтение

stvort commented on library/src/main/java/ru/otus/merets/library/domain/Comment.java:
```diff
@@ -0,0 +1,26 @@
+    @Column(name = "text")
+    private String text;
+
+    @ManyToOne(targetEntity = Book.class, fetch = FetchType.EAGER)
```
А точно нужен Игорь?)

stvort commented on library/src/main/java/ru/otus/merets/library/domain/Book.java:
```diff
@@ -2,17 +2,41 @@
+    @JoinTable(name= "books_authors", joinColumns = @JoinColumn(name="book_id"),inverseJoinColumns = @JoinColumn(name="author_id"))
+    private Set<Author> authors;
+
+    @BatchSize(size = 10)
+    @ManyToMany(targetEntity = Genre.class, fetch = FetchType.EAGER, cascade = CascadeType.ALL)
+    @JoinTable(name= "books_genres", joinColumns = @JoinColumn(name="book_id"),inverseJoinColumns = @JoinColumn(name="genre_id"))
+    private Set<Genre> genres;
```
Возможно тут больше подойдет `@Fetch(FetchMode.SUBSELECT)`

stvort commented on library/src/test/java/ru/otus/merets/library/repository/BookRepositoryJpaTest.java:
```diff
@@ -0,0 +1,110 @@
+    @Test
+    @DisplayName("pass migrations regarding authors")
+    void shouldPassMigrationsViaFlyWayRegardingAuthors() {
+        assertThat(bookRepository.findById(1L).get().getAuthors().size()).isEqualTo(3);
```
Лучше так:
```
assertThat(bookRepository.findById(1L)).isNotEmpty()
                .get()
                .extracting(Book::getAuthors).asList().hasSize(3);
```
